### PR TITLE
Add allow-by-default lint for unit bindings

### LINKED ISSUE #71432: Warn on useless bindings like `let v2 = v1.sort();`
I've seen a lot of beginning Rust programmers make a mistake similar to the program below, which compiles and runs without warning:

```rust
fn main() {
    let mut x = [0,1,2];
    x[0] = 5;
    let y = x.sort();
    println!("{:?}", y);
}
```

The bug in this program is hard for new Rust programmers to spot. The compiler should help them, by warning that `x.sort()` does not return a value.

This could be done with a new attribute `#[must_not_use]`, applied to functions like `sort`. This would be the inverse of `must_use`: It would warn only if the result of the function *is* used. Like `must_use`, it would take an optional argument that provides an additional diagnostic message.

(Bikeshed: `must_not_use` may be a confusing name. Other suggestions welcome.)

Or perhaps any assignment `let foo = expr;` where `expr` has type `()` should cause a warning by default. This would be a more general solution, catching more possible errors but also potentially causing more disruption. The warnings in this case might be less helpful, because they wouldn't include suggestions tailored to specific functions.

### PR DESCRIPTION
**Pull request by jieyouxu**
### Example

```rust
#![warn(unit_bindings)]

macro_rules! owo {
    () => {
        let whats_this = ();
    }
}

fn main() {
    // No warning if user explicitly wrote `()` on either side.
    let expr = ();
    let () = expr;
    let _ = ();

    let _ = expr; //~ WARN binding has unit type
    let pat = expr; //~ WARN binding has unit type
    let _pat = expr; //~ WARN binding has unit type

    // No warning for let bindings with unit type in macro expansions.
    owo!();

    // No warning if user explicitly annotates the unit type on the binding.
    let pat: () = expr;
}
```

outputs

```
warning: binding has unit type `()`
  --> $DIR/unit-bindings.rs:17:5
   |
LL |     let _ = expr;
   |     ^^^^-^^^^^^^^
   |         |
   |         this pattern is inferred to be the unit type `()`
   |
note: the lint level is defined here
  --> $DIR/unit-bindings.rs:3:9
   |
LL | #![warn(unit_bindings)]
   |         ^^^^^^^^^^^^^

warning: binding has unit type `()`
  --> $DIR/unit-bindings.rs:18:5
   |
LL |     let pat = expr;
   |     ^^^^---^^^^^^^^
   |         |
   |         this pattern is inferred to be the unit type `()`

warning: binding has unit type `()`
  --> $DIR/unit-bindings.rs:19:5
   |
LL |     let _pat = expr;
   |     ^^^^----^^^^^^^^
   |         |
   |         this pattern is inferred to be the unit type `()`

warning: 3 warnings emitted
```

This lint is not triggered if any of the following conditions are met:

- The user explicitly annotates the binding with the `()` type.
- The binding is from a macro expansion.
- The user explicitly wrote `let () = init;`
- The user explicitly wrote `let pat = ();`. This is allowed for local lifetimes.

### Known Issue

It is known that this lint can trigger on some proc-macro generated code whose span returns false for `Span::from_expansion` because e.g. the proc-macro simply forwards user code spans, and otherwise don't have distinguishing syntax context compared to non-macro-generated code.

Closes #71432.

### CODE REVIEW

**Comment by est31** [comment]
It's more a lint for bindings of unit type, right? Maybe it should be renamed to `unit_bindings`? Or are there hypothetically other instances where the lint could warn?

I'm wondering because clippy has a let and return lint, something which I am hitting often because I like to make the pattern and don't really like the lint. Still, it would be good to instantly know the difference between the lint this PR is adding and `clippy::let_and_return`.

**Comment by jieyouxu** [comment]
> It's more a lint for bindings of unit type, right? Maybe it should be renamed to `unit_bindings`? Or are there hypothetically other instances where the lint could warn?

Yep, I can rename it `unit_binding`, it's called `useless_binding` now because I couldn't come up with a better name for it.

**Comment by jieyouxu** [comment]
> It's more a lint for bindings of unit type, right? Maybe it should be renamed to `unit_bindings`? Or are there hypothetically other instances where the lint could warn?

Yep, I can rename it `unit_binding`, it's called `useless_binding` now because I couldn't come up with a better name for it.

File: compiler/rustc_lint_defs/src/lib.rs [hunk 1]
```diff
@@ -550,6 +550,9 @@ pub enum BuiltinLintDiagnostics {
         /// The local binding that shadows the glob reexport.
         private_item_span: Span,
     },
+    UnitBindings {
+        span: Span,
+    },
```
Comment (WaffleLapkin) [inline compiler/rustc_lint_defs/src/lib.rs:555]: Is this ever created? I can't find it...

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs [hunk 2]
```diff
@@ -1447,6 +1451,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        decl_ty: Ty<'tcx>,
+        init_ty: Option<Ty<'tcx>>,
+        decl: Declaration<'tcx>,
+    ) {
+        if decl.span.can_be_used_for_suggestions()
+            && let Some(init_ty) = init_ty
+            && self.infcx.can_eq(self.param_env, init_ty, self.tcx.mk_unit())
+            && decl.ty.is_none() // suppress if user explicitly ascribes a type to the pattern
+            && self.infcx.can_eq(self.param_env, decl_ty, self.tcx.mk_unit())
+            && !matches!(decl.pat.kind, hir::PatKind::Wild) // suppress if user writes `let _ = expr;`.
```
Comment (WaffleLapkin) [inline compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1469]: why is this necessary? `let _ = unit()` seems redundant too.

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs [hunk 3]
```diff
@@ -1447,6 +1451,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
             self.diverges.set(previous_diverges);
         }
+
+        self.lint_unit_bindings(decl_ty, init_ty, decl);
+    }
+
+    fn lint_unit_bindings(
```
Comment (WaffleLapkin) [inline compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1458]: Can this use normal linting infrastructure? (if not, why?)

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs (hunk 2 above)
Comment (jieyouxu) [reply compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1469]: > why is this necessary?

Sorry which one? Do you mean "suppress if user explicitly ascribes a type to the pattern"?

> `let _ = unit()` seems redundant too.

I will remove this case.

File: compiler/rustc_lint_defs/src/lib.rs (hunk 1 above)
Comment (jieyouxu) [reply compiler/rustc_lint_defs/src/lib.rs]: (This was for the old bad way of doing lints, but now changed to use the normal way...)

**Review by WaffleLapkin — APPROVED** [review]
The implementation looks good.

**Comment by WaffleLapkin** [comment]
`T-lang` requested a crater run, can you add a temporary commit that makes this lint deny-by-default, so that we can make a crater run and see how much crates have triggered this?

**Comment by jieyouxu** [comment]
> `T-lang` requested a crater run, can you add a temporary commit that makes this lint deny-by-default, so that we can make a crater run and see how much crates have triggered this?

I will make a temporary commit, but I think crater supports `-D unit_bindings` (cf. https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Allowing.20a.20lint.20on.20diesel.20for.20bors.20try/near/328933796). Something like https://github.com/rust-lang/rust/pull/107880#issuecomment-1437505502.

**Comment by jieyouxu** [comment]
I will try to fix these test cases first...:

```
failures:
    [ui] tests/ui/assign-assign.rs
    [ui] tests/ui/associated-types/normalization-debruijn-3.rs
    [ui] tests/ui/associated-type-bounds/trait-alias-impl-trait.rs
    [ui] tests/ui/async-await/issue-64130-4-async-move.rs#drop_tracking
    [ui] tests/ui/async-await/issue-64130-4-async-move.rs#drop_tracking_mir
    [ui] tests/ui/associated-type-bounds/rpit.rs
    [ui] tests/ui/associated-types/issue-54182-1.rs
    [ui] tests/ui/async-await/field-assign.rs#no_drop_tracking
    [ui] tests/ui/async-await/drop-track-field-assign.rs#drop_tracking_mir
    [ui] tests/ui/async-await/field-assign.rs#drop_tracking_mir
    [ui] tests/ui/async-await/drop-track-field-assign.rs#no_drop_tracking
    [ui] tests/ui/async-await/drop-track-field-assign.rs#drop_tracking
    [ui] tests/ui/associated-type-bounds/dyn-rpit-and-let.rs
    [ui] tests/ui/async-await/field-assign.rs#drop_tracking
    [ui] tests/ui/associated-type-bounds/dyn-impl-trait-type.rs
    [ui] tests/ui/binding/nil-pattern.rs
    [ui] tests/ui/async-await/non-trivial-drop.rs#no_drop_tracking
    [ui] tests/ui/async-await/non-trivial-drop.rs#drop_tracking
    [ui] tests/ui/async-await/non-trivial-drop.rs#drop_tracking_mir
    [ui] tests/ui/async-await/issues/issue-55809.rs
    [ui] tests/ui/closures/2229_closure_analysis/run_pass/disjoint-capture-in-same-closure.rs
    [ui] tests/ui/closures/issue-868.rs
    [ui] tests/ui/cfg/cfg_stmt_expr.rs
    [ui] tests/ui/const-generics/issues/issue-70273-assoc-fn.rs
    [ui] tests/ui/const-generics/generic_const_exprs/issue-86710.rs
    [ui] tests/ui/consts/assoc_const_generic_impl.rs
    [ui] tests/ui/consts/const-eval/erroneous-const2.rs
    [ui] tests/ui/consts/const-eval/erroneous-const.rs
    [ui] tests/ui/consts/dangling-alloc-id-ice.rs
    [ui] tests/ui/consts/const-eval/promoted_errors.rs#noopt
    [ui] tests/ui/consts/const-eval/promoted_errors.rs#opt
    [ui] tests/ui/consts/const-eval/promoted_errors.rs#opt_with_overflow_checks
    [ui] tests/ui/consts/const-eval/unwind-abort.rs
    [ui] tests/ui/consts/copy-intrinsic.rs
    [ui] tests/ui/consts/large_const_alloc.rs
    [ui] tests/ui/empty-allocation-rvalue-non-null.rs
    [ui] tests/ui/feature-gates/feature-gate-generic_arg_infer.rs#feature
    [ui] tests/ui/generics/post_monomorphization_error_backtrace.rs
    [ui] tests/ui/higher-ranked/trait-bounds/normalize-under-binder/issue-62529-6.rs
    [ui] tests/ui/issues/issue-11047.rs
    [ui] tests/ui/issues/issue-11740.rs#thir
    [ui] tests/ui/issues/issue-11740.rs#mir
    [ui] tests/ui/issues/issue-17999.rs
    [ui] tests/ui/issues/issue-24353.rs
    [ui] tests/ui/issues/issue-11709.rs
    [ui] tests/ui/for-loop-while/loop-break-value.rs
    [ui] tests/ui/issues/issue-39827.rs
    [ui] tests/ui/issues/issue-23808.rs
    [ui] tests/ui/issues/issue-20644.rs
    [ui] tests/ui/issues/issue-36474.rs
    [ui] tests/ui/issues/issue-48131.rs#mir
    [ui] tests/ui/issues/issue-48131.rs#thir
    [ui] tests/ui/issues/issue-53843.rs
    [ui] tests/ui/lint/lint-removed-cmdline.rs
    [ui] tests/ui/lint/lint-removed.rs
    [ui] tests/ui/lint/lint-renamed-cmdline.rs
    [ui] tests/ui/lint/lint-renamed-allow.rs
    [ui] tests/ui/lint/lint-removed-allow.rs
    [ui] tests/ui/lint/lint-renamed.rs
    [ui] tests/ui/lint/lint-attr-everywhere-early.rs
    [ui] tests/ui/lint/unused/lint-unused-imports.rs
    [ui] tests/ui/lint/unused_import_warning_issue_45268.rs
    [ui] tests/ui/lint/unused/lint-unused-variables.rs
    [ui] tests/ui/lint/lint-attr-everywhere-late.rs
    [ui] tests/ui/intrinsics/panic-uninitialized-zeroed.rs#strict
    [ui] tests/ui/macros/format-args-temporaries.rs
    [ui] tests/ui/match/issue-84434.rs
    [ui] tests/ui/intrinsics/panic-uninitialized-zeroed.rs#default
    [ui] tests/ui/mir/issue-101844.rs
    [ui] tests/ui/mir/issue-80949.rs
    [ui] tests/ui/mir/issue66339.rs
    [ui] tests/ui/never_type/diverging-fallback-unconstrained-return.rs#nofallback
    [ui] tests/ui/never_type/diverging-fallback-unconstrained-return.rs#fallback
    [ui] tests/ui/never_type/diverging-fallback-control-flow.rs#fallback
    [ui] tests/ui/nll/assign-while-to-immutable.rs
    [ui] tests/ui/never_type/diverging-fallback-control-flow.rs#nofallback
    [ui] tests/ui/nll/issue-54943-3.rs
    [ui] tests/ui/no-warn-on-field-replace-issue-34101.rs
    [ui] tests/ui/moves/pin-mut-reborrow-infer-var-issue-107419.rs
    [ui] tests/ui/polymorphization/drop_shims/transitive.rs
    [ui] tests/ui/reachable/expr_return.rs
    [ui] tests/ui/polymorphization/drop_shims/simple.rs
    [ui] tests/ui/raw-ref-op/unusual_locations.rs
    [ui] tests/ui/return-nil.rs
    [ui] tests/ui/resolve/resolve-pseudo-shadowing.rs
    [ui] tests/ui/crate-method-reexport-grrrrrrr.rs
    [ui] tests/ui/rfcs/rfc-2421-unreserve-pure-offsetof-sizeof-alignof/offsetof-alignof-sizeof-pure-can-be-used-as-idents.rs
    [ui] tests/ui/rust-2018/dyn-keyword.rs
    [ui] tests/ui/rfcs/rfc-2565-param-attrs/param-attrs-allowed.rs
    [ui] tests/ui/rfcs/rfc-2565-param-attrs/param-attrs-cfg.rs
    [ui] tests/ui/svh/svh-use-trait.rs
    [ui] tests/ui/macros/macro-comma-support-rpass.rs#core
    [ui] tests/ui/traits/bound/generic_trait.rs
    [ui] tests/ui/tuple/index-float.rs
    [ui] tests/ui/try-block/issue-45124.rs
    [ui] tests/ui/unboxed-closures/unboxed-closures-zero-args.rs
    [ui] tests/ui/macros/macro-comma-support-rpass.rs#std
    [ui] tests/ui/weird-exprs.rs
    [ui] tests/ui/resolve/extern-prelude.rs
    [ui] tests/ui/proc-macro/issue-81555.rs
    [ui] tests/ui/lint/unused/lint-unused-extern-crate.rs
    [ui] tests/ui/closures/deeply-nested_closures.rs

test result: FAILED. 6 passed; 102 failed; 15024 ignored; 0 measured; 0 filtered out; finished in 2.35s

Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Build completed unsuccessfully in 0:00:04
```

**Comment by danielhenrymantilla** [comment]
Related to this change, the following limitation:

  - issue: [let_unit_value should be allowed when I explicitly write "let () = ..."](https://github.com/rust-lang/rust-clippy/issues/9048)
  - PR _in progress_: [Don't lint let_unit_value when () is explicit](https://github.com/rust-lang/rust-clippy/pull/10844)

As in, while I agree that _implicit_ bindings of `()` have a rather high likelihood of being an oversight, and thus, a mistake/logic bug (**which makes the lint, and this PR, very useful!**), there are more rare but nonetheless legitimate cases where a `()` value is bound, but in a properly annotated way to acknowledge this fact (either by using the `()` pattern, or a `: ()` type annotation). In such cases, this lint firing is a false-positive, and thus, an annoyance. So this is something to be aware of ⚠️ 

I think it will just be a matter of waiting for https://github.com/rust-lang/rust-clippy/pull/10844 to be merged, and we will be good to go! 🙂

**Comment by jieyouxu** [comment]
Opened #112549 to adjust multiple UI tests to help reduce the amount of test failures.

**Comment by jieyouxu** [comment]
I updated the lint to not fire in the following cases:

- The user explicitly wrote `let () = init;`.
- The user explicitly wrote `let pat = ();`.

**Comment by jieyouxu** [comment]
Maybe the `unit_bindings` lint should also not fire when the binding is `let _`?

**Comment by WaffleLapkin** [comment]
@jieyouxu I think it makes sense to lint for `let _`, unsure if there are any cases where you'd want to ignore the unit like that (i.e. without asserting the type like `let () = ` or `let _: () = ` do)

**Comment by jieyouxu** [comment]
> Oh, in this case I think you can drop the commit and we can start a perf run with the flag

I updated two more UI tests and dropped the commit, it shouldn't fail in UI tests now (maybe in other places, but I haven't done a full build yet).

**Comment by WaffleLapkin** [comment]
@crater check start=mester#2ca8d358e55bc56755b597ea96b557232ef8bc86 end=try#f6cbccd7d7fb2aee12357229da0c0b36b1ad87b6+rustflags=-Dunit_bindings

**Comment by WaffleLapkin** [comment]
@craterbot check start=master#2ca8d358e55bc56755b597ea96b557232ef8bc86 end=try#f6cbccd7d7fb2aee12357229da0c0b36b1ad87b6+rustflags=-Dunit_bindings

**Comment by WaffleLapkin** [comment]
@craterbot check start=master#2ca8d358e55bc56755b597ea96b557232ef8bc86 end=try#aa6a960545be254a71d7a833b7485e29e974b433+rustflags=-Dunit_bindings

**Comment by jieyouxu** [comment]
# Crater Run Analysis

## Analysis Method

Analysis is achieved by the small script at <https://github.com/jieyouxu/unit-bindings-analyze-crater-report>.

## Results

There are **3100** regressed crates which were reported by crater to be caused by
`unit_bindings`. These crates produced a total of **17792** error messages which 
are actually due to `unit_bindings`.

The error messages can be classified into **11** categories:

1. Lints fired on procedural-macro-annotated functions: **220**
2. Lints fired on enumset macro: **2**
3. Lints fired on content mixed with special macros: **24**
4. Lints fired on py03-related macros: **19**
5. Lints fired on special UI macros: **4**
6. Lints fired on unknown crate (can't find source): **2**
7. Lints fired on `let mut? _ = expr`: **6154**
8. Lints fired on `let mut? var_name = expr`: **5961**
9. Lints fired on `let mut? _underscore_var_name = expr`: **4588**
10. Lints fired on JSX-like exprs `<JSXFragment />`: **752**
11. Lints fired on wildcard-assign exprs `_ = expr`: **66**

The lint currently is still incorrectly fired on some (proc-)macros 
(**1021/17792** total error messages) which introduce local bindings with spans 
that return `false` for `Span::from_expansion()`.

The lint correctly fires and produces **16769/17792** of the error messages.

## Lists of Crates and Short Examples for Some of the Incorrectly Fired Categories

### Procedural-macro-annotated functions

These are typically misfired on Rocket-like procedural macro annotated functions.

#### Example

Lint fires on

```rs
#[post("/message", data = "<form>")]
fn post(form: Form<Message>,queue: &State<Sender<Message>>) {
    let _res = queue.send(form.into_inner());
}
```

```
[INFO] [stdout] error: binding has unit type `()`
[INFO] [stdout]   --> src/main.rs:22:4
[INFO] [stdout]    |
[INFO] [stdout] 22 | fn post(form: Form<Message>,queue: &State<Sender<Message>>) {
[INFO] [stdout]    |    ^^^^ this pattern is inferred to be the unit type `()`
[INFO] [stdout]    |
[INFO] [stdout]    = note: requested on the command line with `-D unit-bindings`
```

#### Crates

<details>
    <summary>Show crates</summary>

```
"Abhishekjena93.chat-app.084e4a325e8a0d95a8affa5a4d6ef1d2e3d73ee3",
"AleardoKandiah.rust-app.a39ae916de97d3a933f6f7e9c33cdfeeb33ea9e5",
"Al-khali.rust_chat_app.a9373c9e1cb04ee91004ef88a3a8d224e23ff3b3",
"Anderson0xFF.TradeEngine_API.923bd444b3dbc4ebce4471f9c2d385478e6e6ffa",
"ArtiomTr.boxer.a7f0f72dc45cb736ec8302a84985763fb35ddcba",
"BawnX.rocket-first.308cd7c06f3fa147c3b9ce7f392f475978607ac9",
"BroGamer4256.divamodarchive.4a283bcd1854b01597003a7b24b6a4eacd2923a2",
"Dheatly23.godot-wasm.f913b071d12aa2b14d055191a38d8049ed662b93",
"Eliasin.sysc4806project.7faf1acf5033fdfae43638c11a45f87d67cf20bd",
"JesseAbram.rocket_testing_issue.ec4baac3bb96b6f04ab72f975dfdc6673b45b4bb",
"KaratsubaLabs.onigiri-server.f65db5dc7ad73638fc8a74c348cdc83b5324e8e9",
"Lunarequest.lunatree.40442a27853728cf861751af5b19f8173152332d",
"Misterio77.SCC0541-Lab-BD-Projeto.06b4b27a80bf408fa9155f17c0df3918d58716dc",
"NeoLight1010.spots-api.12f64a39d2cbc8d05c42bde7afcb01b85d39729c",
"Nexlson.Mahjong-Backend.593b2430248b6dfa4d0e85323a92027dc786c462",
"NolValue.atomic.c9d4e0183a28f9c62da939b9d28c2ed0963fd216",
"Pigolitsyn.rust-server.4dbf049be83a7606710719c1735041938da9aa14",
"PygmalionAI.collect_data.2f5ffc11b184ea119140a9537843ae535bf89fe0",
"R0CKB0TT0M.plexhue.f2741e10853bb3070fcd02bc20c3d1b319b72a5c",
"Rushmore75.pulse_remote.006b7c5b702994ae72956cad2cf00f995a387a35",
"SerggioPizzella.beam.72d07a4450b21945f3e016c27bc0c2b3286fa1b2",
"acmutd.jury.4ce478239ff56f60fa163f0a39d0d59907121ab4",
"aiguy110.webhook2websocket.fbec5736936502dbbde8cdc6c2636695a73a7727",
"cassaundra.uwupaste.112189a73504076aa479bec537bef43d8c8c8209",
"chriamue.nft-login.d7b12cc14fc949a21af6cc33daf14f8987c6c9cd",
"clonejo.clubstatusd.b5ab4e1919eced214db4cc1c4b0b5170f404676e",
"cooscoos.RustBucket.67734afa6b3a9007101aa2aad0ffee5d3f4d3288",
"cwhakes.cwh-basic.45f6cadc6902620eb3f46d98df186ac33cb5b8f1",
"digitalillusion.remote_colonies.a14ec3ae3467317df46f2b97c88571db6059325b",
"divinerapier.rust-examples.d45e78a2006170efefb841887e3d0bebf93d316d",
"duaifzn.linebot.4ec52ed102f7a21c9adf2488313fd0d914c068de",
"eframework-0.1.0",
"fazekaszs.protonation-states-backend.8ff2ac3e2d6c53ae0fe208ae787489f6f2dbc60a",
"gilengel.proc-server.4f92a1e972c26c0957a729f8be9140cceedbe37b",
"hamidrezakp.kanvas.ed52c2b83d68483a15059b3ac8306c95db07933c",
"hashed-io.bdk-services.6efcc7c952b74195ca985bb851421d13beb04ee3",
"heyjul.BeSTS_API.fec01e52ab1675961b005bdc38fd0e82981cfabc",
"id-contact.auth-test.c467d2863a9ca83d40787f239f6d17f0d32c8362",
"jbfp.videocaster.3020e4d01fdd65883af91497411b7d7782c3b982",
"jusvait.HelplyBackend.348ce0fe2eea681d9825360d487489a35d65f804",
"kennycallado.PFC-server.4bac75490841cd5f4427724f52cd878780f33c06",
"kpkym.koe-server.94eab9e57e81390e7ed8b74c6e04ee4229919c1e",
"memochou1993.shortener-rust-api.6c214a1d41feda75d459a3cd6c801603ff45f430",
"mkmik.ocipfs.476d377a42adbe6e33766629a3786858963c079d",
"mtso.rabbithole.8e15470e5aa7ec22186a8814f5c27d67e0cda04d",
"mvl-at.keg.0728900261db9e812917b6b2d57c4546f605fc9e",
"oakpr.leaderboard.2860cd21a3cc653fba81e454a5bfcdd995cc42f4",
"offdroid.swmn.8186d268da9b173247854e0ca003cd605b8e0b2d",
"pettermk.kiddybank-api.febc38123e124349899bc11992eac863d6b56ee9",
"pixelmonaskarion.messenger.e5aa63f292c8a0da66e490cc8f5c5608bed1d462",
"pseudobabble.rocket-hello.eed7aedc225b17f28ae817b32c8a7005d91fa184",
"quartz-server-0.3.0",
"racetrack-0.0.2",
"rishiad.rust_auth_server.753ba3e08068db0c980ad2d8cbce34b6832da253",
"rivalq.Yew-Rocket-Todo.81f2bb22dc888954012c67e1c9294b951b355da2",
"rob9315.ping-mc.b5891023d830d38b0de7ec3dc80ea7f9c318f6fe",
"rocket-0.5.0-rc.3",
"rocket_csrf-0.3.0",
"rocket_modules-0.1.1",
"roliveiravictor.demo-rust-learning.17c6d10d7e28e528fcf048650b8f62a9c8ab40a8",
"rstropek.RustyRockets.27ff816d3612b162916fff85e6e61d160c64b128",
"shaku_rocket-0.7.0-rc.1",
"stakwork.sphinx-swarm.81ae1bbd0f793a164f78d66ee610f97b09b18b6b",
"tandem_http_server-0.3.0",
"tannaurus.Grandma.50497d763b3b56e55508cb165fd461f8519ed756",
"the-auction-games.account-api.fe8149f89fb03519c3a10f87da42e3a2da3b37b0",
"tweedegolf.tguard.61e6e3be8427d548e369c301d727698e794d07b8",
"yaaawww.keyman.e5173b2cd233530dd4056bf6e2440ba765616650",
"yoehwan.rust_todo_server.c0ed7ae2ce437faf506fcd619ff378e6e296731c",
"yunusp.rental.a8f65bacff43a9e9d317d50070083cc52911bcf6",
"zhangyi921.is-color-orange.b1d9a1947a4e46adde787cfd149e293c9eb31150",
```
</details>

### Enumset macro

#### Example

```
[INFO] [stdout] error: binding has unit type `()`
[INFO] [stdout]   --> src/objects/socket.rs:26:11
[INFO] [stdout]    |
[INFO] [stdout] 26 | #[enumset(serialize_as_list)]
[INFO] [stdout]    |           ^^^^^^^^^^^^^^^^^ this pattern is inferred to be the unit type `()`
[INFO] [stdout]    |
[INFO] [stdout]    = note: requested on the command line with `-D unit-bindings`
```

#### List of crates

```
"BlankParenthesis.pxls-rs.d4ec12b7c6ffdaa457098c621a4522f245ecfc45"
```

### Content mixed with special macros

These typically have heavy procedural macro content with custom syntax.

#### Example

```
[INFO] [stdout] error: binding has unit type `()`
[INFO] [stdout]   --> src/parser.rs:20:19
[INFO] [stdout]    |
[INFO] [stdout] 20 |       %syntax_error { match token {
[INFO] [stdout]    |  ___________________^
[INFO] [stdout] 21 | |         Some(token) => Err(extra.0.syntax_error(extra.1.clone(), format!("unexpected token: {:?}", token))),
[INFO] [stdout] 22 | |         None => Err(extra.0.syntax_error(extra.1.clone(), "unexpected end of input".into())),
[INFO] [stdout] 23 | |     }};
[INFO] [stdout]    | |______^ this pattern is inferred to be the unit type `()`
[INFO] [stdout]    |
[INFO] [stdout]    = note: requested on the command line with `-D unit-bindings`
```

#### List of crates

<details>
    <summary>Show crates</summary>

```
"intercom-0.4.0",
"lets_expect-0.5.1",
"linyinfeng.rspg.b431fb00f34ff81ae5425e14494747df0c9e297b",
"openrr-plugin-0.1.0",
"phaneron-plugin-0.1.2",
"pomelo-0.1.5",
"smt2parser-0.6.1",
```
</details>

### PyO3-related macros

#### Example

```
[INFO] [stdout] error: binding has unit type `()`
[INFO] [stdout]    --> src/lib.rs:121:7
[INFO] [stdout]     |
[INFO] [stdout] 121 |     #[args(
[INFO] [stdout]     |       ^^^^ this pattern is inferred to be the unit type `()`
[INFO] [stdout]     |
[INFO] [stdout]     = note: requested on the command line with `-D unit-bindings`
```

#### List of crates

<details>
    <summary>Show crates</summary>

```
"messense.lingua-py.3d97ee130f5b4954f89575be1fec90c730686be6",
"messense.rjieba-py.63536c2f69282584431bfddd00d234ad27ae0e1c",
"mozilla.pyo3-parsepatch.9bc6c2d5c2150818f79e300d301848a57e29b2c0",
"pyo3-0.19.0",
```
</details>

### Special UI macros

(This should technically go with JSX-like macros.)

#### Example

```
[INFO] [stdout] error: binding has unit type `()`
[INFO] [stdout]   --> www/src/main.rs:21:18
[INFO] [stdout]    |
[INFO] [stdout] 21 |                 <GameUI<Mark,Grid,TicTacToe>/>
[INFO] [stdout]    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^ this pattern is inferred to be the unit type `()`
[INFO] [stdout]    |
[INFO] [stdout]    = note: requested on the command line with `-D unit-bindings`
```

#### List of crates

```
"prestonmlangford.mcts.20e683ca78398bffbc7ee2cccc3ec6040bd4d210"
```

## Unknown Crates

I couldn't find the relevant source for `musli-tests-0.0.3`.

## Analysis Log

<details>
    <summary>Analysis Program Debug Output</summary>

```
total_error_messages: 17792
proc macro fn crates removed; remaining error messages: 17572
enumset crates removed; remaining error messages: 17570
DEBUG unit_bindings_analyze_crater_report: remaining_crates = 3025
special macros crates removed; remaining error messages: 17546
py03 macros crates removed; remaining error messages: 17527
special ui macro crates removed; remaining error messages: 17523
idk crates removed; remaining error messages: 17521
let_underscore_messages: 6154
DEBUG unit_bindings_analyze_crater_report: remaining_crates_1.len() = 1679
let_var_messages: 5961
DEBUG unit_bindings_analyze_crater_report: remaining_crates_2.len() = 772
let_underscore_var_messages: 4588
DEBUG unit_bindings_analyze_crater_report: remaining_crates_3.len() = 242
jsx_like_messages: 752
DEBUG unit_bindings_analyze_crater_report: remaining_crates_4.len() = 26
underscore_assign_messages: 66
DEBUG unit_bindings_analyze_crater_report: remaining_crates_5.len() = 0
```

</details>

**Comment by WaffleLapkin** [comment]
@jieyouxu I'm a bit lost, what are the next steps here?

Do you want to fix the macro issue?

> The lint currently is still **incorrectly** fired on some (proc-)macros
> (**1021/17792** total error messages) which introduce local bindings with spans
> that return false for `Span::from_expansion()`.

**Comment by jieyouxu** [comment]
> Do you want to fix the macro issue?

I'm not sure I know *how* to fix the macro issues: the problem is that these constructed macros don't have distinguishting context from non-macro code AFAICT?

**Comment by WaffleLapkin** [comment]
Nominating for T-lang discussion, as they are usually the ones approving lints.

The current lint behavior is described in the PR description. It's designed in such a way that there shouldn't be many false positives (is this the right term? I'm always confusing them...) — as using an explicit unit type, a `()` pattern or a `()` expression will silence the lint.

The only false positive really possible is the case of proc macro using a user provided span for the unit binding. There isn't much to do here and this seems to be an issue with the proc macro implementation, not the lint. See this comment for details: https://github.com/rust-lang/rust/pull/112380#issuecomment-1612576955.

The **question(s) for the lang team**: do we want to include such lint in the compiler? Is the current behavior the one we want?

**Comment by joshtriplett** [comment]
As discussed in today's lang meeting:

@rfcbot merge

**Comment by joshtriplett** [comment]
(Huh. Race condition with rfcbot?)

**Comment by WaffleLapkin** [comment]
Writing this down, so this is not lost. In today's lang meeting we've discussed that the plan should probably be:
1. Merge this lint as allow-by-default (after FCP is complete)
2. Ping proc macro authors / submit patches / announce (https://github.com/rust-lang/rust/labels/relnotes, maybe inside rust blog) and wait for a "long enough time"
3. Bump this lint to warn-by-default

**Comment by nikomatsakis** [comment]
@WaffleLapkin 

> false positives (is this the right term? I'm always confusing them...) 

minor note: once I read somewhere about how a less confusing terminology would be to say "false warnings" or "missed warnings", and that's what I try to say no, precisely because I always am confused myself.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by nikomatsakis** [comment]
Unnominating as there are no further questions to be answered by lang team

@rustbot labels -I-lang-nominated

File: compiler/rustc_lint/src/unit_bindings.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+declare_lint_pass!(UnitBindings => [UNIT_BINDINGS]);
+
+impl<'tcx> LateLintPass<'tcx> for UnitBindings {
+    fn check_local(&mut self, cx: &crate::LateContext<'tcx>, local: &'tcx hir::Local<'tcx>) {
+        if !local.span.from_expansion()
+            && let Some(tyck_results) = cx.maybe_typeck_results()
+            && let Some(init) = local.init
+            && let init_ty = tyck_results.expr_ty(init)
+            && let local_ty = tyck_results.node_type(local.hir_id)
```
Comment (WaffleLapkin) [inline compiler/rustc_lint/src/unit_bindings.rs:54]: I understand that that is a much more rare case, but should we maybe inspect the pattern instead to find bindings inside? example:
```rust
fn tuple<A, B>(a: A, b: B) -> (A, B) { (a, b) }

let (unit, _) = tuple((), 12);
```

**Comment by WaffleLapkin** [comment]
@nikomatsakis thanks for the tip, I'll try to use this terminology in the future =)

**Comment by scottmcm** [comment]
This looks amazing.  I agree that unit bindings are usually a mistake, I'm glad to see the exceptions for `let () = `/`let pointless: () =` if someone *really* wants it, and landing allow and upping to warn later makes good sense.

@rfcbot reviewed

**Comment by safinaskar** [comment]
Is it true that clippy's lint let_unit_value will become not needed when this is merged?

**Comment by WaffleLapkin** [comment]
@safinaskar I believe so, yes.

cc @rust-lang/clippy then (I was not aware that clippy had a lint that is almost the same)

**Comment by Alexendoo** [comment]
I think it makes sense to keep `let_unit_value` while `unit_bindings` is allowed by default, we can mark it as uplifted when it becomes a warning

**Comment by WaffleLapkin** [comment]
@Alexendoo I agree, makes sense 👍🏻

File: compiler/rustc_lint/src/unit_bindings.rs (hunk 4 above)
Comment (jieyouxu) [reply compiler/rustc_lint/src/unit_bindings.rs:54]: Sorry for the late reply. I don't really have a strong opinion on this. I would expect that this kind of code pattern to be rather rare?

**Comment by est31** [comment]
the build failure is legit, likely needs a rebase plus `x fmt` (rustfmt now formats let chains). @rustbot author
